Extended interface for PANonlinearFormExtention and its gradient [tmopt-split-api]

vladotomov opened the pull request:
Extracted from the #1484 (tmopt) branch.

One of the three parallel PRs that split #1484 

This is an API PR that extends the `NonlinearFormExtension` interface @jandrej and `OperatorJacobiSmoother` @barker29. It doesn't contain anything that tests the new code, as all that is in #1484. We should test there before merging to `master`, or just merge into #1484.

One big question for me was whether `NonlinearForm::Setup, NonlinearForm::SetupGrad` should have an argument `const Vector &x`. To compute any data at quad points for a nonlinear function, the `x` argument is needed. Otherwise one can only do size allocations, as in #1484, while all quad point computations will be in the Integrator’s `MultPA` and `MultGradPA`. On the other hand, adding the argument is confusing, because one implicitly assumes that `Setup(x)` and the following `MultPA(x)` use the same `x` (same for `SetupGradient(x)`, `GetGradient(x)`). I didn’t add the argument to the interface for now, as #1484 works without it, but I’m curious what others think.

The same can be said about `NonlinearFormIntegrator::AssemblePA()` with `x` argument; also `PANonlinearFormExtension::Assemble` and `PANonlinearFormExtension::AssembleGradient`.

<!--GHEX{"id":1731,"author":"vladotomov","editor":"v-dobrev","reviewers":["camierjs","tzanio"],"assignment":"2020-09-10T17:47:35-07:00","approval":"2020-11-10T19:49:49.129Z","merge":"2021-04-08T20:40:01.967Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1731](https://github.com/mfem/mfem/pull/1731) | @vladotomov | @v-dobrev | @camierjs + @tzanio | 09/10/20 | 11/10/20 | 04/08/21 | |
<!--ELBATXEHG-->

v-dobrev commented:
Additional reviews are also welcome.

cc: @mfem/reviewers

rcarson3 commented on fem/nonlininteg.hpp:
```diff
@@ -88,6 +97,17 @@ class NonlinearFormIntegrator
        called. */
    virtual void AddMultPA(const Vector &x, Vector &y) const;
 
+   /// Method for partially assembled gradient action.
+   /** All arguments are E-vector. This method can be called only after the
+       method AssembleGradPA() has been called. */
+   virtual void AddMultGradPA(const Vector &g,
+                              const Vector &x, Vector &y) const;
+
+   /// Method for computing the diagonal of the gradient with partial assmebly.
+   /** All arguments are E-vector. This method can be called only after the
+       method AssembleGradPA() has been called. */
+   virtual void AssembleGradDiagonalPA(const Vector &x, Vector &diag) const;
```
So, I've implemented in the above in my own branch of MFEM for ExaConstit as seen  [here](https://github.com/rcarson3/mfem/blob/exaconstit-dev/fem/nonlininteg.hpp#L102-L107). I guess I'm a little curious about what is the ```const Vector &g``` in ```AddMultGradPa``` and ```const Vector &x``` in ```AssembleGradDiagonalPA``` are supposed to be just because I at least didn't require these for my own implementations.

vladotomov replied:
I described the arguments in the comments, please take a look. The extra arguments are related to the comment in the PR description -- the PA-related `Setup` and `SetupGrad` of the `NonlinearForm` don't have an argument `x`, thus one needs to pass it it to integrators (by `g` in this case). Do you think they should have `x` as an argument?

rcarson3 replied:
Okay, I see what you're talking about now.

> Do you think that `Setup` and `SetupGrad` should have `x` as an argument?

So, based on what I've done in `ExaConstit` I do feel that they should have a vector argument to them, which could be named something else to differentiate it from the vectors going into the `MultPA` and `MultGradPA` functions. The current `Setup` is a bit constraining in this manner and for example forces me to do something like this in my `NonlinearForm`-based class's [Mult](https://github.com/LLNL/ExaConstit/blob/exaconstit-dev/src/mechanics_operator.cpp#L279-L286).

One other change I would make to  `Setup` and `SetupGrad` is also have them be virtual functions on the `NonlinearForm` class. I know within `ExaConstit` my `Setup` is actually both the `Setup` and `SetupGrad` due to how my problems are formulated both of these are best done at the same time. I also make use of the `Setup` phase for all of my different assembly operations, since they all share the info obtained from it.

rcarson3 commented:
Overall, a lot of the work you've done inside of the ```NonlinearFormExtension``` appears to be similar to what I'd done in ExaConstit for my own if you will for [my own](https://github.com/LLNL/ExaConstit/blob/exaconstit-dev/src/mechanics_operator_ext.cpp) parallel version of a ```NonlinearFormExtension``` like class. Which, I guess shouldn't be too surprising since I based quite a bit of it on the ```BilinearForm``` implementations.  However, I did make a few changes to mine which I feel brings it closer to how things are done within the ```BilinearFormExtention``` class where the ```Mult``` represents the action of the gradient/matrix on a vector rather than what's in the ```NonlinearFormExtension``` where ```Mult``` was made to be something closer to actions of a ```LinearFormIntegrator```. 

I can't say I currently have too many questions about it. Although, I am intrigued by your ```GetGradient``` method where it appears to construct a new ```Gradient``` object every time its called.  I'm wondering if this really necessary, since it appears to me at least that you're having to constantly create the space for the new vectors.

vladotomov commented:
> However, I did make a few changes to mine which I feel brings it closer to how things are done within the `BilinearFormExtention` class where the `Mult` represents the action of the gradient/matrix on a vector rather than what's in the `NonlinearFormExtension` where `Mult` was made to be something closer to actions of a `LinearFormIntegrator`.

I don't understand this, could you be more specific what can be improved?

> `GetGradient` method where it appears to construct a new `Gradient` object every time its called.

I changed this, thanks.

rcarson3 commented:
> > However, I did make a few changes to mine which I feel brings it closer to how things are done within the `BilinearFormExtention` class where the `Mult` represents the action of the gradient/matrix on a vector rather than what's in the `NonlinearFormExtension` where `Mult` was made to be something closer to actions of a `LinearFormIntegrator`.
> 
> I don't understand this, could you be more specific what can be improved?

So part of this goes back to the original PR, and how I felt we should've be careful with the naming of things to make sure what we're doing in the `BilinearFormExtension` is consistent with what's being done in the `NonlinearFormExtension`. Since it appeared to me, we could eventually make `NonlinearFormExtension` the parent class for `BilinearFormExtension` because they share so much of the same framework underneath. Since, the `Mult` in both classes represented different actions that weren't consistent with one another which could end up being confusing to users of the code. However after some thought on this, I do think your design could still work if the `BilinearFormExtension` was eventually refactored to be based on the `NonlinearFormExtension`.

As far as things that could be improved for this class, I would say you could probably reuse the `xe` and `ye` that already live on the `PANonlinearFormExtension` within its `Gradient` class. It would be a nice little memory saving optimization. Also, you could maybe create the `Gradient` at the same time the class is constructed. Then in the `GetGradient` function you could get rid of the if else statement and just have `dynamic_cast<PANonlinearFormExtension::Gradient *>(Grad.Ptr())->ReInit(x);` Although, I guess if we change the `SetupGrad` to be able to take in that `Vector x` then you could further simplify `GetGradient` to just be `return *Grad.Ptr();`.

vladotomov commented:
> based on what I've done in ExaConstit I do feel that they should have a vector argument to them

Thanks, lets see what others think too.

> One other change I would make to `Setup` and `SetupGradient` is also have them be virtual functions on the NonlinearForm class

Why? I think these are called only by the end user, so if he inherits `NonlinearForm`, and wants to do something different, he can simply create and call a different function. Similarly, `BilinearForm::Assemble` is not `virtual`.

> we could eventually make `NonlinearFormExtension` the parent class for `BilinearFormExtension`.
> I do think your design could still work if the BilinearFormExtension was eventually refactored to be based on the NonlinearFormExtension.

I'm not sure what would be the point of having this connection. `NonlinearForm` and `BilinearForm` are not connected in this way.

> Also, you could maybe create the `Gradient` at the same time the class is constructed.

There are use cases where a `NonlinearForm` is used without the need to compute its gradient.

> I would say you could probably reuse the xe and ye that already live on the PANonlinearFormExtension within its Gradient class.

As above, the `Gradient` may not exist.

rcarson3 commented:
> > One other change I would make to `Setup` and `SetupGradient` is also have them be virtual functions on the NonlinearForm class
> 
> Why? I think these are called only by the end user, so if he inherits `NonlinearForm`, and wants to do something different, he can simply create and call a different function. Similarly, `BilinearForm::Assemble` is not `virtual`.

It makes sense a lot more sense when you think about children classes of a `NonlinearForm` that have much more complicated `Setup` phases being used in something like the `NewtonRaphson` solver class.  You might have the NR solver case just take in a `NonlinearForm` that then calls `NonlinearForm::Setup` before doing the `NonlinearForm::Mult`. It's here that you'd want `Setup` to be virtual. Also, I must have over looked it earlier but `NonlinearForm::Setup()` is already virtual in master and probably for similar reasons.

> > we could eventually make `NonlinearFormExtension` the parent class for `BilinearFormExtension`.
> > I do think your design could still work if the BilinearFormExtension was eventually refactored to be based on the NonlinearFormExtension.
> 
> I'm not sure what would be the point of having this connection. `NonlinearForm` and `BilinearForm` are not connected in this way.

They may not be connected that way, but based on my experience a lot of the code already written in the `BilinearFormExtension` class can work with the `NonlinearFormExtension`without any real modifications. So, what I really wish here is for both of these to have a base class that both of these can just build off of their common code patterns.

> > Also, you could maybe create the `Gradient` at the same time the class is constructed.
> 
> There are use cases where a `NonlinearForm` is used without the need to compute its gradient.

I know, but it was just a thought on ways you could simplify the code path in the `GetGradient` call.

> > I would say you could probably reuse the xe and ye that already live on the PANonlinearFormExtension within its Gradient class.
> 
> As above, the `Gradient` may not exist.

My thoughts here would be for `xe` and `ye` to exist on the `PANonlinearFormExtension` class, and then the `Gradient` class would just have pointers pointed to these `Vectors`.

tzanio reviewed (approved):
This looks good to me, thanks @vladotomov and @camierjs.

I don't think `NonlinearForm::Setup` should have a `Vector` argument -- that doesn't make sense for a nonlinear operator because any computations will be invalidated by a different output. The Setup is still useful though for allocating memory and preparing the prolongation operators.

For `NonlinearForm::SetupGrad` it depends if we want to do partial assembly or matrix-free. For PA we do need a `Vector` argument, for MF we don't. The PA version is probably preferable in this case.

tzanio commented:
I forgot to mention: maybe the `README` should be updated?

camierjs commented on fem/nonlininteg.cpp:
```diff
@@ -28,12 +35,32 @@ void NonlinearFormIntegrator::AssemblePA(const FiniteElementSpace &,
+void NonlinearFormIntegrator::AssembleGradDiagonalPA(const mfem::Vector &g,
+                                                     mfem::Vector &diag) const
+{
+   mfem_error ("NonlinearFormIntegrator::AssembleDiagonalPA(...)\n"
```
`AssembleGradDiagonalPA` vs `AssembleDiagonalPA`

camierjs replied:
`done`

camierjs commented:
LGTM

vladotomov commented:
I've updated the branch with the latest changes:
- Added `x` argument to `SetupGradient` and down the chain.
- Fixed errors that came up during testing -- the `Gradient` operator of the `PAExtension` had to work on ldofs.

We can update the CHANGELOG when we merge `tmopt` and have a working example of nonlinear Newton solve with PA.

camierjs commented:
LGTM

camierjs commented:
LGTM

camierjs commented:
LGTM

rcarson3 commented on fem/nonlinearform_ext.hpp:
```diff
@@ -17,28 +17,75 @@
+   NonlinearFormExtension(const NonlinearForm*);
+
+   /// Assemble at the AssemblyLevel of the subclass.
+   virtual void Assemble() = 0;
```
This would technically be a breaking change if introduced. I'm don't think it'll affect anyone, since this is largely used internally within MFEM, but it is something we should be mindful about.

rcarson3 replied:
Discussed in a MFEM meeting, and it was decided that this was an appropriate change due it correcting the previous name which incorrectly suggested that the method should only be used for partial assembly rather than for PA, EA, or FA.

rcarson3 commented on fem/nonlinearform_ext.hpp:
```diff
@@ -17,28 +17,75 @@
+      const Operator *elemR;
+      const FiniteElementSpace &fes;
+      const Array<NonlinearFormIntegrator*> &dnfi;
+      mutable Vector ge, xe, ye, ze;
```
I still believe that two of these vectors could use the storage from the vectors `xe` and `ye` that are already created when `PANonlinearFormExtension` is made. However, you could do this in a separate PR as further optimizations to the class.

rcarson3 reviewed (approved):
I know my approval isn't required... but it looks fine to me. It's fairly similar to what I'd already done in ExaConstit and my branch of mfem for exaconstit. Also, I don't see it clashing in the few places we went down different paths for how things should be done, so I don't have a problem seeing it get merged in.

tzanio commented:
thank you @rcarson3 !

v-dobrev commented:
Merged in `next` for testing...

v-dobrev commented:
LGTM

v-dobrev commented on fem/nonlinearform_ext.hpp:
```diff
@@ -17,28 +17,75 @@
+
+   /// Assemble at the AssemblyLevel of the subclass.
+   virtual void Assemble() = 0;
+   /// Assemble gradient data at the AssemblyLevel of the subclass, for the
+   /// state @a x, which is assumed to be a ldof Vector.
+   virtual void AssembleGradient(const Vector &x) = 0;
+
+   /// Assumes that @a x is a ldof Vector.
+   virtual Operator &GetGradient(const Vector &x) const = 0;
```
Do we need both of these methods? `AssembleGradient` seems redundant because `GetGradient` also takes `x` as input.

vladotomov replied:
This is related to having separate `NonlinearForm::SetupGradient` and `NonlinearForm::GetGradient` for the PA case. I thought we decided to have the option to do quadrature computations separately.

v-dobrev commented:
I created a branch (https://github.com/mfem/mfem/tree/tmopt-split-api-updates) and PR (#2018 toward this PR) with a set of proposed updates for this PR.

Please review these proposed changes and approve them (PR #2018) if you are okay to merge them here.

ping: @vladotomov, @camierjs, @tzanio, @rcarson3

tzanio commented:
Please resolve conflicts and update `CHANGELOG`

camierjs commented:
LGTM

v-dobrev commented:
Testing in https://github.com/mfem/mfem/pull/1822

adrienbernede commented:
LGTM

adrienbernede commented:
Any new commit discards the previous LGTM comment.

v-dobrev commented:
Re-merged in `next` for testing.

v-dobrev commented:
LGTM
